tools/release/release_deb_monorepo.py

```diff
@@ -0,0 +1,422 @@
+# as published by the Free Software Foundation.
+#
+# Checkbox is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Checkbox.  If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import json
+import logging
+import os
+import shutil
+import subprocess
+
+from pathlib import Path
+
+
+class ConsoleFormatter(logging.Formatter):
+
+    """Custom Logging Formatter to ease copy paste of commands."""
+
+    def format(self, record):
+        fmt = '%(message)s'
+        if record.levelno == logging.ERROR:
+            fmt = "%(levelname)-8s %(message)s"
+        result = logging.Formatter(fmt).format(record)
+        return result
+
+
+# create logger
+logger = logging.getLogger('release')
+logger.setLevel(logging.DEBUG)
+# create file handler which logs even debug messages
+fh = logging.FileHandler('release.log')
+fh.setLevel(logging.DEBUG)
+# create console handler with a higher log level
+ch = logging.StreamHandler()
+ch.setLevel(logging.INFO)
+# create formatter and add it to the handlers
+fh_formatter = logging.Formatter('%(asctime)-15s %(levelname)-8s %(message)s')
+fh.setFormatter(fh_formatter)
+ch.setFormatter(ConsoleFormatter())
+# add the handlers to the logger
+logger.addHandler(fh)
+logger.addHandler(ch)
+
+
+def run(*args, **kwargs):
+    """wrapper for subprocess.run."""
+    try:
+        return subprocess.run(
+            *args, **kwargs,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError as e:
+        logger.error('{}\n{}'.format(e, e.output.decode()))
+        raise SystemExit(1)
+
+
+class Release:
+
+    def __init__(self, args) -> None:
+        self.args = args
+        self.settings = {
+            "dry_run": False,
+            "mode": "testing",
+            "root": None,
+            "projects": {}
+        }
+        self._checkpoint = Path('.checkpoint.json')
+
+    def _save_checkpoint(self):
+        with open(self._checkpoint, "w") as f:
+            json.dump(self.settings, f, indent=4, sort_keys=True)
+
+    def _load_checkpoint(self):
+        self.settings = json.load(open(self._checkpoint))
+
+    def clone(self):
+        config = json.load(open(self.args.config))
+        mode = config["mode"]
+        logger.info("".center(80, '#'))
+        logger.info("# Parsing local repo...")
+        logger.info("".center(80, '#'))
+        for path, dirs, files in os.walk('.'):
+            if "debian" in dirs:
+                project_path = str(Path(*Path(path).parts))
+                project_root = '.'
+                project_name = str(project_path).replace('s/', '-')
+                if not config[project_name]:
+                    self.settings["projects"][project_name] = {
```

Comment:
A subjective stylistic point made as a general grumble: to me the whole logic would come out easier to follow if there was actually no mutable `self.settings`, but there were free functions / class methods that took `settings` as an argument, and returned a copy of the modified `settings` as a return value, for the next party to then mutate.